UPGRADE.md

```diff
@@ -0,0 +1,152 @@
+
+The SDK extracts the error details from the response and stores them in the error object, along with the request ID and the HTTP status code.
+
+`AbstractNylasSdkError` is used for errors returned by the SDK. Right now there's only one type of error we return, and that's a `NylasSdkTimeoutError` which is thrown when a request times out.
+
+Putting it all together, the following example code shows how to make a request to create a new Event and handle any errors that may occur:
+
+```ruby
+require 'nylas'
+
+nylas = Nylas::API.new(
+  api_key: "NYLAS_APP_KEY",
+)
+
+begin
+  # Build the create event request
+  create_event_request = nylas.events.create(
+    identifier: "GRANT_ID",
+    query_params: {
+      calendar_id: "CALENDAR_ID", # A calendar ID is required
+    },
+    request_body: {
+      when: {
+        start_time: 1686765600,
+        end_time: 1686769200,
+      }, # A "When" type is required
+      title: "My Event", # Title is optional
+      description: "My event description", # Description is optional
+      location: "My event location", # Location is optional
+    }
+  )
+rescue Nylas::NylasApiError => e
+  # Handle the error
+  puts e.message
+  puts e.request_id
+  puts e.status_code
+rescue Nylas::NylasSdkTimeoutError => e
+  # Handle the error
+  puts e.message
+  puts e.url
+end
+```
+
+## Authentication
+
+The SDK's authentication methods reflect [the methods available in the new Nylas API v3](https://developer.nylas.com/docs/developer-guide/v3-authentication/). While you can only create and manage your application's connectors (formerly called integrations) in the dashboard, you can manage almost everything else directly from the SDK. This includes managing grants, redirect URIs, OAuth tokens, and authenticating your users.
+
+There are two main methods to focus on when authenticating users to your application. The first is the `Auth#urlForOAuth2` method, which returns the URL that you redirect your users to in order to authenticate them using Nylas' OAuth 2.0 implementation.
+
+The second is the `Auth#exchangeCodeForToken` method. Use this method to exchange the code Nylas returned from the authentication redirect for an access token from the OAuth provider. Nylas's response to this request includes both the access token, and information about the grant that was created.  You don't _need_ to use the `grant_id` to make requests. Instead, you can use the authenticated email address directly as the identifier for the account. If you prefer to use the `grant_id`, you can extract it from the `CodeExchangeResponse` object and use that instead.
+
+The following code shows how to authenticate a user into a Nylas application:
+
+```ruby
+require 'nylas'
+
+nylas = Nylas::API.new(
+  api_key: "NYLAS_APP_KEY",
+)
+
+# Build the URL for authentication
+auth_url = nylas.auth.url_for_oauth2(
+  client_id: "CLIENT_ID",
+  redirect_uri: "REDIRECT_URI",
```

Comment:
Hey @mrashed-dev is this the Nylas `redirect_uri` or the customer application's `callback_uri`?  We should be super clear about those since we're doing the renames. :)

Reply:
As per our conversation, we're holding off on mentioning the readme in the SDKs until the API team deploys the changes to staging and docs are updated. Once the contract is cemented, we'll update the SDK and these readme docs, thanks!